Custom partition assignment

mmaslankaprv opened the pull request:
## Cover letter

Added support for manual partition assignment in `CreateTopics` API. The manual partition assignment allow user to specify which nodes should be members of a partition replica set. When custom partition assignment is specified number of partitions and replication factor are derived from assignments itself. 

### Implementation details

Custom partition assignment is passed by the client when creating topics. We do not differentiate manually assigned partitions from the one that are assigned automatically. The custom assignment portion of the `CreateTopics` request is used only during the assignment process itself. This approach have some advantages since the whole mechanics related with automatic partition rebalancing, nodes decommissioning and partition movement in general doesn't have to special case those manually assigned partitions. 

Release note: 

- Added support for manual partition assignment in `CreateTopics` API
  - Currently custom partition assignment is only used during initial partition allocation
  - Nodes decommissioning and automatic partition re-balancing may change the partition assignment

dotnwat commented on src/v/cluster/cluster_utils.cc:
```diff
@@ -185,6 +185,21 @@ std::vector<topic_result> create_topic_results(
+    std::vector<topic_result> results;
+    results.reserve(requests.size());
+    std::transform(
+      std::cbegin(requests),
+      std::cend(requests),
```
nit: for non-templated code (generic over container type) I'm not sure that std::begin offers any advantage

dotnwat commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -305,29 +306,75 @@ make_error_result(const model::topic_namespace& tp_ns, std::error_code ec) {
+        return errc::topic_invalid_replication_factor;
+    }
+
+    if (assignable_config.has_custom_assignment()) {
+        for (auto& custom : assignable_config.custom_assignments) {
+            if (
+              static_cast<int16_t>(custom.replicas.size())
+              != assignable_config.cfg.replication_factor) {
+                return errc::topic_invalid_replication_factor;
+            }
+        }
+    }
```
do we cover validation where custom assignment does use the same node?

mmaslankaprv replied:
yes, there is even a test for this, partition allocation will fail

dotnwat commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -305,29 +306,75 @@ make_error_result(const model::topic_namespace& tp_ns, std::error_code ec) {
         partition_allocator::shard,
-        [t_cfg](partition_allocator& al) {
-            return al.allocate(make_allocation_request(t_cfg));
+        [assignable_config](partition_allocator& al) mutable {
+            return al.allocate(make_allocation_request(assignable_config));
```
`mutable` looks unnecessary here (or you wanted a `std::move`)

dotnwat commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -470,7 +517,8 @@ ss::future<std::vector<topic_result>> topics_frontend::autocreate_topics(
     if (leader == _self) {
         return create_topics(
-          std::move(topics), model::timeout_clock::now() + timeout);
+          without_custom_assignments(topics),
```
std::move(topics)?

dotnwat commented on tests/rptest/tests/custom_topic_assignment_test.py:
```diff
@@ -0,0 +1,100 @@
+
+        def replica_matches():
+            replicas_per_partition = {}
+            for p in list(rpk.describe_topic(name)):
```
if `describe_topic` returns a `map` object the iteration here should unpack that without needing to materialize the whole list with `list()`

dotnwat reviewed (approved):
lgtm.

can you add some detail to the PR cover letter about the overall strategy? For example, it looks like (and maybe I'm missing something) that the custom assignment specified when creating topics is used to controller the allocator, but that the assignment information isn't preserved along with the topic metadata. that is, if partitions were to rebalance or move later in the life time of the cluster the originally requested assignments wouldn't be known?

how does this behave for partition rebalancing / node decommission etc..?

mmaslankaprv commented:
> lgtm.
> 
> 1. can you add some detail to the PR cover letter about the overall strategy? For example, it looks like (and maybe I'm missing something) that the custom assignment specified when creating topics is used to controller the allocator, but that the assignment information isn't preserved along with the topic metadata. that is, if partitions were to rebalance or move later in the life time of the cluster the originally requested assignments wouldn't be known?
> 2. how does this behave for partition rebalancing / node decommission etc..?
> 3. were any on disk data structures change? it didn't look like it, which is sort of what prompted question (1) above.

I've added a section answering questions 1,2 and 3 in cover letter. However, maybe we should mark partition as manually assigned and do not move it during automatic partition rebalancing. Node decommissioning would have to stay the same as there is no choice on what to do with partition (we would require manual intervention otherwise), what do you think ?

dotnwat commented:
> > lgtm.
> > 
> > 1. can you add some detail to the PR cover letter about the overall strategy? For example, it looks like (and maybe I'm missing something) that the custom assignment specified when creating topics is used to controller the allocator, but that the assignment information isn't preserved along with the topic metadata. that is, if partitions were to rebalance or move later in the life time of the cluster the originally requested assignments wouldn't be known?
> > 2. how does this behave for partition rebalancing / node decommission etc..?
> > 3. were any on disk data structures change? it didn't look like it, which is sort of what prompted question (1) above.
> 
> I've added a section answering questions 1,2 and 3 in cover letter. However, maybe we should mark partition as manually assigned and do not move it during automatic partition rebalancing. Node decommissioning would have to stay the same as there is no choice on what to do with partition (we would require manual intervention otherwise), what do you think ?

I can't find any information on what the expected behavior is in Kafka. But, the API is really hard inflexible. Instead, partition assignments should have been based on a declarative specification so we didn't have to deal with the ambiguity of dealing with decommissioning nodes, rebalancing, etc...

What if we (1) add some detail to the release notes about the current behavior of the feature and (2) merge this and then (3) add additional, permanent affinity information later if requested. Adding the concept of fixed affinity and then dealing with all of the fall out of how that interacts with rebalancing seems like a bit much right now.

mmaslankaprv commented:
> > > lgtm.
> > > 
> > > 1. can you add some detail to the PR cover letter about the overall strategy? For example, it looks like (and maybe I'm missing something) that the custom assignment specified when creating topics is used to controller the allocator, but that the assignment information isn't preserved along with the topic metadata. that is, if partitions were to rebalance or move later in the life time of the cluster the originally requested assignments wouldn't be known?
> > > 2. how does this behave for partition rebalancing / node decommission etc..?
> > > 3. were any on disk data structures change? it didn't look like it, which is sort of what prompted question (1) above.
> > 
> > 
> > I've added a section answering questions 1,2 and 3 in cover letter. However, maybe we should mark partition as manually assigned and do not move it during automatic partition rebalancing. Node decommissioning would have to stay the same as there is no choice on what to do with partition (we would require manual intervention otherwise), what do you think ?
> 
> I can't find any information on what the expected behavior is in Kafka. The API is really inflexible. Instead, I wonder why kafka didn't make partition assignments be based on a declarative specification so we didn't have to deal with the ambiguity of dealing with decommissioning nodes, rebalancing, etc... oh well...
> 
> What if we (1) add some detail to the release notes about the current behavior of the feature and (2) merge this and then (3) add additional, permanent affinity information later if requested. Adding the concept of fixed affinity and then dealing with all of the fall out of how that interacts with rebalancing seems like a bit much right now.

I've added information to Release notes on how we handle manually assigned partitions during parititon rebalancing.

dotnwat commented:
> 

Thanks for updating. This seems like a good start. We can figure out what to do then persistent affinity is needed.
